### PR TITLE
feat: Adding a tutorial for deploying UPF in DPDK mode

### PR DESCRIPTION
**Pull request by gatici**
# Description

The implementation of the DPDK support in the UPF has been introduced in https://github.com/canonical/sdcore-upf-operator/pull/57. 
However, documentation of this feature was missing.
Hence, this PR adds a tutorial about using UPF in DPDK mode.

# Checklist:

- [x] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that validate the behaviour of the software
- [ ] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have bumped the version of the library

### CODE REVIEW

File: docs/tutorials/using_dpdk.md [hunk 1]
```diff
@@ -0,0 +1,632 @@
...
+```
+
+Apply the network configuration:
+
+```shell
+sudo chmod 600 /etc/netplan/50-cloud-init.yaml
+sudo netplan apply
+```
+
+### Create the MACVLAN bridges for `enp7s0` and `enp8s0`
```
Comment (markbeierl) [inline docs/tutorials/using_dpdk.md:264]: This is not needed. When we unbind using `driverctl`, these interfaces go away. The `diverctl` and `enable_unsafe_noiommu_mode` could perhaps be put instead into `/etc/rc.local`

Comment (gatici) [reply docs/tutorials/using_dpdk.md]: Create the MACVLAN bridges part is removed and other configurations are persisted using `/etc/rc.local`.

Comment (gatici) [reply docs/tutorials/using_dpdk.md]: Done

File: docs/tutorials/using_dpdk.md [hunk 2]
```diff
@@ -0,0 +1,618 @@
...
+- CPU that supports AVX2, RDRAND and PDPE1GB instructions (Intel Haswell, AMD Excavator or equivalent)
+- 8 cores
+- 16 GB RAM
+- 50 GiB disk
+
+## 1. Prepare the Host machine
+
+### Enable HugePages
+
+As a `root` user, update the Grub to enable 2 * 1Gi HugePages in the host machine. Then, the host is gracefully rebooted to activate the settings.
```
Comment (markbeierl) [inline docs/tutorials/using_dpdk.md:20]:
```suggestion
As a `root` user, update Grub to enable 2 * 1Gi HugePages in the host machine. This is followed by a gracefully reboot to activate the settings.
```

File: docs/tutorials/using_dpdk.md [hunk 3]
```diff
@@ -0,0 +1,618 @@
...
+The following IP networks will be used for User Plane function deployment:
+
+| Name         | Subnet        | Gateway IP |
+|--------------|---------------|------------|
+| `management` | 10.201.0.0/24 | 10.201.0.1 |
+| `access`     | 10.202.0.0/24 | 10.202.0.1 |
+| `core`       | 10.203.0.0/24 | 10.203.0.1 |
+| `ran`        | 10.204.0.0/24 | 10.204.0.1 |
+
+On the host machine, create local network bridges to be used by LXD by adding below configuration under `/etc/netplan/99-sdcore-networks.yaml`.
```
Comment (markbeierl) [inline docs/tutorials/using_dpdk.md:61]: I'm not sure this howto should cover the creation of the network interfaces. Perhaps this should be its own howto?

Comment (markbeierl) [reply docs/tutorials/using_dpdk.md:61]: Whoops, sorry, I thought I was reading the howto PR, not the tutorial. Ignore this comment.

File: docs/tutorials/using_dpdk.md (hunk 2 above)
Comment (gatici) [reply docs/tutorials/using_dpdk.md]: Changed. Thanks.

File: docs/.wordlist.txt [hunk 4]
```diff
@@ -17,8 +19,10 @@ Charmhub
 CharmHub
 CIDR
 CNI
+cni
 conf
 config
+ConfigMap
```
Comment (markbeierl) [inline docs/.wordlist.txt:25]: Isn't this on L27 already?

File: docs/tutorials/using_dpdk.md [hunk 5]
```diff
@@ -0,0 +1,618 @@
+# Using DPDK
```
Comment (markbeierl) [inline docs/tutorials/using_dpdk.md:1]: Maybe instead call this

# Accelerated Networking

Using DPDK just seems to vague

File: docs/tutorials/using_dpdk.md [hunk 6]
```diff
@@ -0,0 +1,618 @@
...
+To complete this tutorial, you will need one virtual machine with access to the networks as follows:
+
+| Machine                              | CPUs | RAM | Disk | Networks                       |
+|--------------------------------------|------|-----|------|--------------------------------|
+| User Plane Kubernetes Cluster        | 4    | 12g | 20g  | `management`, `access`, `core` |
+
+Create VM with Multipass:
+
+```console
+multipass launch -c 4 -m 12G -d 20G -n user-plane  --network mgmt-br --network core-br --network access-br jammy
```
Comment (markbeierl) [inline docs/tutorials/using_dpdk.md:165]: We reserved the huge pages on the host, but we are not launching the VM using those huge pages. Looking at the multipass documentation, I cannot see a method of launching the VM to use that huge page. If MP cannot, then we should remove the huge page configuration from the host machine configuration as it is locking the 2G RAM that nothing can use.

File: docs/tutorials/using_dpdk.md (hunk 5 above)
Comment (gatici) [reply docs/tutorials/using_dpdk.md:1]: Yes, I like your suggestion.

File: docs/tutorials/using_dpdk.md (hunk 6 above)
Comment (markbeierl) [reply docs/tutorials/using_dpdk.md:165]: I just did a check on my laptop. No huge pages configured
```
mark@mark-laptop:~$ grep Huge /proc/meminfo
AnonHugePages:         0 kB
ShmemHugePages:   831488 kB
FileHugePages:         0 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:               0 kB
```

Launched a MP vm, added 1g huge pages to grub, and it has huge pages after reboot
```
ubuntu@user-plane:~$ grep Huge /proc/meminfo 
AnonHugePages:         0 kB
ShmemHugePages:        0 kB
FileHugePages:         0 kB
HugePages_Total:       2
HugePages_Free:        2
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:    1048576 kB
Hugetlb:         2097152 kB
```

So it is very unlikely that multipass is actually using the host.

File: docs/.wordlist.txt (hunk 4 above)
Comment (gatici) [reply docs/.wordlist.txt:25]: Yes, removed.

File: docs/tutorials/using_dpdk.md (hunk 5 above)
Comment (gatici) [reply docs/tutorials/using_dpdk.md:1]: Done

File: docs/tutorials/using_dpdk.md (hunk 6 above)
Comment (gatici) [reply docs/tutorials/using_dpdk.md]: I tested it again and you are right. Enabling hugepages in the host part is unnecessary. This part is removed.

**Review by markbeierl — APPROVED** [review]

**Pull request merged**